storage: fix data race in MVCCPredicateDeleteRange

stevendanna opened the pull request:
This fixes two bugs in MVCCPredicateDeleteRange that would result in incorrect logical operations being published to rangefeeds.

When publishing a logical operation, the caller indicates whether the buffers provide for the Key and EndKey of the operation are safe for concurrent use by setting a `Safe` boolean field in the published operation.

MVCCPredicateDeleteRange calls either MVCCDeleteRangeUsingTombstone or mvccPutInternal, both of which publish logical opts with Safe set to true.

Unfortunatelely, the code in MVCCPredicateDeleteRange _did_ invalidate the keys passed to these functions.

As a result, it was possible for rangefeeds to observe:

1. Range keys with incorrect bounds,
2. Missed pointed deletions events,
3. Unnecessarily duplicated point deletions, and
4. Point deletions that never occurred (but were part of range deletions).

Additionally, since we expect to be able to read the values for a logical operation back out of storage, rangefeeds may also encounter errors as they failed to read back the value for an erroneous events.

Fixes #112733

Epic: none

Release note(bug fix): Fix a bug that would result in incorrect physical replication results for users of the private preview physical replication feature.

stevendanna committed:
storage: fix data race in MVCCPredicateDeleteRange

This fixes two bugs in MVCCPredicateDeleteRange that would result in
incorrect logical operations being published to rangefeeds.

When publishing a logical operation, the caller indicates whether the
buffers provide for the Key and EndKey of the operation are safe for
concurrent use by setting a `Safe` boolean field in the published
operation.

MVCCPredicateDeleteRange calls either MVCCDeleteRangeUsingTombstone or
mvccPutInternal, both of which publish logical opts with Safe set to
true.

Unfortunatelely, the code in MVCCPredicateDeleteRange _did_ invalidate
the keys passed to these functions.

As a result, it was possible for rangefeeds to observe:

1. Range keys with incorrect bounds,
2. Missed pointed deletions events,
3. Unnecessarily duplicated point deletions, and
4. Point deletions that never occurred (but were part of range deletions).

Additionally, since we expect to be able to read the values for a
logical operation back out of storage, rangefeeds may also encounter
errors as they failed to read back the value for an erroneous events.

Fixes #112733

Epic: none

Release note(bug fix): Fix a bug that would result in incorrect
physical replication results for users of the private preview physical
replication feature.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/112845)
<!-- Reviewable:end -->

stevendanna commented on pkg/storage/mvcc.go:
```diff
@@ -3777,10 +3780,11 @@ func MVCCPredicateDeleteRange(
-				// buffer's current position. No data corruption occurs because the
-				// buffer is flushed up to runSize.
-				buf[runSize] = append(buf[runSize][:0], runEnd...)
+				// TODO(ssd): It would be nice to avoid unnecessary allocations
+				// here. We copy the end key because when mvccInternalPut publishes
+				// the logical operation related to the delete, it currenty assumes
+				// that the key will not be subsequently modified.
+				buf = append(buf, runEnd.Clone())
```
@msbutler I think you are familiar with this code.  Let me know if you have an opinion here on a good allocation strategy.   Maybe instead of this, perhaps we keep the overwritting, but then allocate a whole new buf anytime we flush point deletions. That way we don't pay allocation costs in the case that we were able to generate range deletes. Do we have a benchmark for this function anywhere?

msbutler replied:
I'm unsure which strategy wins. If we see a pattern of many switches between point key runs and range key runs, then I think your current PR wins, while a pattern of fewer switches would prefer your alternative approach. I didn't write a benchmark function anywhere, and it seems would should add one that looks like https://github.com/cockroachdb/cockroach/blob/2b18f4ffc260dca938ba0265217c4483ec6064c7/pkg/storage/bench_pebble_test.go#L432

stevendanna committed:
storage: reduce number of allocation in MVCCPredicateDeleteRange

Here, we use bufalloc to reduce the overall number of
allocations. This substantially reduces overall allocations in our
benchmarks at the expense of a 3-5% runtime regression.

Epic: none

Release note: None

stevendanna commented:
@msbutler I played a bit with different allocation strategies but so far they've mostly been a mixed bag in your benchmark.  This change does create a big increase in allocations:

```
Without this change, we see a pretty large increase in the number of
allocations in our curren benchmark:

                                                                            │ master.txt  │    change_without_bufalloc.txt     │
                                                                            │   sec/op    │   sec/op     vs base               │
MVCCDeleteRangeWithPredicate_Pebble/streakBound=10/rangeKeyThreshold=64-10     4.144 ± 2%    4.182 ± 2%       ~ (p=0.221 n=20)
MVCCDeleteRangeWithPredicate_Pebble/streakBound=100/rangeKeyThreshold=64-10    1.808 ± 1%    1.791 ± 1%       ~ (p=0.211 n=20)
MVCCDeleteRangeWithPredicate_Pebble/streakBound=200/rangeKeyThreshold=64-10   574.4m ± 1%   594.8m ± 3%  +3.55% (p=0.000 n=20)
MVCCDeleteRangeWithPredicate_Pebble/streakBound=500/rangeKeyThreshold=64-10   60.50m ± 3%   62.94m ± 2%  +4.03% (p=0.000 n=20)
geomean                                                                       714.3m        727.6m       +1.87%

                                                                            │ master.txt  │     change_without_bufalloc.txt     │
                                                                            │  allocs/op  │  allocs/op   vs base                │
MVCCDeleteRangeWithPredicate_Pebble/streakBound=10/rangeKeyThreshold=64-10    2.188k ± 1%   3.424k ± 1%  +56.47% (p=0.000 n=20)
MVCCDeleteRangeWithPredicate_Pebble/streakBound=100/rangeKeyThreshold=64-10   1.695k ± 2%   2.518k ± 1%  +48.58% (p=0.000 n=20)
MVCCDeleteRangeWithPredicate_Pebble/streakBound=200/rangeKeyThreshold=64-10   1.668k ± 1%   2.026k ± 2%  +21.46% (p=0.000 n=20)
MVCCDeleteRangeWithPredicate_Pebble/streakBound=500/rangeKeyThreshold=64-10   2.026k ± 1%   2.163k ± 1%   +6.71% (p=0.000 n=20)
geomean                                                                       1.882k        2.479k       +31.75%
```

Using our `bufalloc` package, I can bring the allocation cost down substantially, but at the cost of the runtime of the benchmark:


```
                                                                            │ master.txt  │      change_with_bufalloc.txt      │
                                                                            │   sec/op    │   sec/op     vs base               │
MVCCDeleteRangeWithPredicate_Pebble/streakBound=10/rangeKeyThreshold=64-10     4.144 ± 2%    4.437 ± 3%  +7.08% (p=0.000 n=20)
MVCCDeleteRangeWithPredicate_Pebble/streakBound=100/rangeKeyThreshold=64-10    1.808 ± 1%    1.921 ± 2%  +6.27% (p=0.000 n=20)
MVCCDeleteRangeWithPredicate_Pebble/streakBound=200/rangeKeyThreshold=64-10   574.4m ± 1%   607.2m ± 2%  +5.72% (p=0.000 n=20)
MVCCDeleteRangeWithPredicate_Pebble/streakBound=500/rangeKeyThreshold=64-10   60.50m ± 3%   65.47m ± 4%  +8.20% (p=0.000 n=20)
geomean                                                                       714.3m        763.0m       +6.81%

                                                                            │ master.txt  │      change_with_bufalloc.txt       │
                                                                            │  allocs/op  │  allocs/op   vs base                │
MVCCDeleteRangeWithPredicate_Pebble/streakBound=10/rangeKeyThreshold=64-10    2.188k ± 1%   2.538k ± 1%  +15.97% (p=0.000 n=20)
MVCCDeleteRangeWithPredicate_Pebble/streakBound=100/rangeKeyThreshold=64-10   1.695k ± 2%   1.687k ± 3%        ~ (p=0.189 n=20)
MVCCDeleteRangeWithPredicate_Pebble/streakBound=200/rangeKeyThreshold=64-10   1.668k ± 1%   1.639k ± 1%   -1.77% (p=0.002 n=20)
MVCCDeleteRangeWithPredicate_Pebble/streakBound=500/rangeKeyThreshold=64-10   2.026k ± 1%   1.970k ± 2%   -2.76% (p=0.000 n=20)
geomean                                                                       1.882k        1.928k        +2.47%
```

Perhaps this is worth the trade-off in the real world.

msbutler commented on pkg/storage/mvcc.go:
```diff
@@ -3595,7 +3596,12 @@ func MVCCPredicateDeleteRange(
+	// for. We copy the keys using keyAlloc, truncating keyAlloc
+	// if we don't send the point deletes and creating a new
+	// keyAlloc if we do send the point deletes.
+	var keyAlloc bufalloc.ByteAllocator
```
Also, I understand we can truncate the byteAllocator if we issue a range tombstone -- those point keys will never get used again. 

Why don't you truncate the buffer after the point key flush as well? We truncate the buffer in both cases.

stevendanna replied:
Here "truncate" means "possibly reuse the underlying capacity.  But in the case of issuing the point deletes, that underlying capacity is being used for the Key's for the deletes we've issued and if we reuse the memory then we might end up issuing logical operations for the wrong key because of the memory overwrite.

stevendanna replied:
> it looks like the ByteAllocator has no unit tests and is currently only used to generate workloads. Before using this package in 23.2/1 for C2C, we ought to unit test this package. I am happy to help with this effort in a separate PR.

It is used in a few more places than that:

- `pkg/kv/kvserver/rangefeed/catchup_scan.go`
- `pkg/kv/kvserver/gc/gc.go`
- `pkg/sql/rowcontainer/disk_row_container.go`
- A few places in changefeeds

I agree some unit tests for this package would be prudent, but we definitely aren't the first production user of this.

msbutler replied:
> But in the case of issuing the point deletes, that underlying capacity is being used for the Key's for the deletes we've issued and if we reuse the memory then we might end up issuing logical operations for the wrong key because of the memory overwrite.

aha. what I initially missed is why it's ok to _always_ truncate the buffer via `buf := buf[:0]`, regardless of a point or range key flush: the `buf` stores  _pointers_ to `roachpb.keys` since they keys are actually arrays of bytes. In `mvccPutInternal( buf[i],...)`, we're passing a _pointer_ to a key stored in `keyAlloc`.  It's fine to truncate the buf since `mvccPutInternal( buf[i],...)` has the right address to the key to delete. 

I think it may be worth explaining in the docstring why you only truncate after range deletes and why its safe to always truncate buf.

> [byteAllocator] is used in a few more places than that:
LOL, if I wait _several_ seconds, goland eventually does find all those references.

msbutler commented on pkg/storage/testdata/mvcc_histories/delete_range_predicate:
```diff
@@ -47,7 +47,7 @@ meta: "i"/0,0 -> txn={id=00000001 key=/Min iso=Serializable pri=0.00000000 epo=0
 data: "i"/7.000000000,0 -> /BYTES/i7
 
 # Writing next to or above point keys and tombstones should work.
-run stats ok
```
curious: with your log-ops flag, do any of these unit tests fail without your change?

stevendanna replied:
Yup!  The reason I needed to add the extra test case was specifically to get a test that would produce incorrect range key operations.  Some of the existing tests already produce incorrect point deletion operations.

msbutler reviewed (approved):
LGTM! I only have clarification questions and a testing follow up task I can take on.

stevendanna commented:
bors r=msbutler

stevendanna commented:
blathers backport 23.2

stevendanna commented:
blathers backport 23.1
